File: ports/python/src/unicode_python/security/crypto/bip39_canonical.py
"""BIP-39 wordlist layer for Bip39Canonical.

Mirrors sections 5-6 of ``Unicode.Security.Crypto.Bip39Canonical``: splitting a
canonical-form codepoint sequence into words at ``U+0020`` boundaries, and
membership / language resolution against the ten BIP-39 wordlists (English,
Japanese, Korean, Spanish, Chinese Simplified, Chinese Traditional, French,
Italian, Czech, Portuguese). Words and wordlist entries are compared as
codepoint sequences, matching the kernel-visible ``wordlistCps`` tables in
``Unicode.Generated.BIP39``.

The canonicalisation pipeline (NFKD -> toLower -> collapse whitespace -> trim)
and the top-level ``detect`` compose on top of this layer; ``toLower`` depends
on the casing subsystem and lands with it.
"""

from dataclasses import dataclass, field
from enum import Enum
from pathlib import Path

from ..casing import Locale, to_lower
from ..identity.ucd import to_nfkd

_DATA_DIR = Path(__file__).resolve().parent.parent.parent / "data" / "bip39"


class Language(Enum):
    """The ten BIP-39 wordlist languages. The ``value`` is the vendored
    wordlist filename stem."""

    ENGLISH = "english"
    JAPANESE = "japanese"
    KOREAN = "korean"
    SPANISH = "spanish"
    CHINESE_SIMPLIFIED = "chinese_simplified"
    CHINESE_TRADITIONAL = "chinese_traditional"
    FRENCH = "french"
    ITALIAN = "italian"
    CZECH = "czech"
    PORTUGUESE = "portuguese"


# Declaration order, matching ``Unicode.Generated.BIP39.allLanguages``. English
# is first, so ``unique_language`` resolves an input that several wordlists
# could cover to English exactly as the Lean ``findSome?`` over ``allLanguages``
# does.
ALL_LANGUAGES = [
    Language.ENGLISH,
    Language.JAPANESE,
    Language.KOREAN,
    Language.SPANISH,
    Language.CHINESE_SIMPLIFIED,
    Language.CHINESE_TRADITIONAL,
    Language.FRENCH,
    Language.ITALIAN,
    Language.CZECH,
    Language.PORTUGUESE,
]

_wordlist_cache: dict[Language, frozenset[tuple[int, ...]]] = {}


def _wordlist_cps(lang: Language) -> frozenset[tuple[int, ...]]:
    """The language's 2,048 words as codepoint tuples, parsed once from the
    vendored line-separated wordlist and memoised."""
    cached = _wordlist_cache.get(lang)
    if cached is not None:
        return cached
    text = (_DATA_DIR / f"{lang.value}.txt").read_text(encoding="utf-8")
    entries = frozenset(
        tuple(ord(ch) for ch in line) for line in text.split("\n") if line != ""
    )
    _wordlist_cache[lang] = entries
    return entries


def split_words(canonical: list[int]) -> list[list[int]]:
    """Split a canonical-form codepoint sequence into words at ``U+0020``
    boundaries; empty runs between separators contribute no word."""
    words: list[list[int]] = []
    current: list[int] = []
    for cp in canonical:
        if cp == 0x0020:
            if current:
                words.append(current)
                current = []
        else:
            current.append(cp)
    if current:
        words.append(current)
    return words


def is_in_wordlist(lang: Language, word: list[int]) -> bool:
    """True iff ``word`` (as codepoints) appears in ``lang``'s wordlist."""
    return tuple(word) in _wordlist_cps(lang)


def wordlists_containing(word: list[int]) -> list[Language]:
    """Every language whose wordlist contains ``word``; empty if none do."""
    return [lang for lang in ALL_LANGUAGES if is_in_wordlist(lang, word)]


def unique_language(words: list[list[int]]) -> Language | None:
    """The first language whose wordlist covers every word in ``words``, else
    ``None``. Empty ``words`` returns English (every predicate holds vacuously),
    matching the Lean ``findSome?`` over ``allLanguages``."""
    for lang in ALL_LANGUAGES:
        if all(is_in_wordlist(lang, word) for word in words):
            return lang
    return None


# ─────────────────────────────────────────────────────────────────────
# Canonicalisation pipeline: NFKD -> toLower(default) -> collapse -> trim.
# ─────────────────────────────────────────────────────────────────────


def is_bip39_whitespace(cp: int) -> bool:
    """The two BIP-39 word separators: U+0020 SPACE and U+3000 IDEOGRAPHIC
    SPACE (Japanese mnemonics)."""
    return cp == 0x0020 or cp == 0x3000


def collapse_whitespace_to_single(cps: list[int]) -> list[int]:
    """Replace every maximal run of BIP-39 whitespace with a single U+0020."""
    out: list[int] = []
    in_ws = False
    for cp in cps:
        if is_bip39_whitespace(cp):
            if not in_ws:
                out.append(0x0020)
            in_ws = True
        else:
            out.append(cp)
            in_ws = False
    return out


def trim_leading_trailing(cps: list[int]) -> list[int]:
    """Strip leading and trailing U+0020 (after whitespace has been collapsed)."""
    start = 0
    end = len(cps)
    while start < end and cps[start] == 0x0020:
        start += 1
    while end > start and cps[end - 1] == 0x0020:
        end -= 1
    return cps[start:end]


def bip39_canonical(cps: list[int]) -> list[int]:
    """The BIP-39 canonical form: NFKD, then default-locale lowercase, then
    whitespace collapse, then leading/trailing trim (spec order)."""
    return trim_leading_trailing(
        collapse_whitespace_to_single(to_lower(Locale.DEFAULT, to_nfkd(cps)))
    )


# ─────────────────────────────────────────────────────────────────────
# Hazard probes (per-priority position finders).
# ─────────────────────────────────────────────────────────────────────


def count_trailing_whitespace(cps: list[int]) -> int:
    count = 0
    for cp in reversed(cps):
        if is_bip39_whitespace(cp):
            count += 1
        else:
            break
    return count


def first_uppercase_pos(cps: list[int]) -> int | None:
    for index, cp in enumerate(cps):
        if 0x41 <= cp <= 0x5A:
            return index
    return None


def first_whitespace_run_pos(cps: list[int]) -> int | None:
    """First position of a leading or consecutive BIP-39 whitespace run; single
    internal separators do not fire."""
    for index, cp in enumerate(cps):
        if is_bip39_whitespace(cp):
            if index == 0:
                return index
            if index + 1 < len(cps) and is_bip39_whitespace(cps[index + 1]):
                return index
    return None


def first_array_divergence(a: list[int], b: list[int]) -> int | None:
    """First position at which ``a`` and ``b`` differ (in element, or one ends);
    ``None`` when identical."""
    for index in range(min(len(a), len(b))):
        if a[index] != b[index]:
            return index
    if len(a) != len(b):
        return min(len(a), len(b))
    return None


# ─────────────────────────────────────────────────────────────────────
# Top-level detection.
# ─────────────────────────────────────────────────────────────────────


@dataclass(frozen=True, slots=True)
class Classification:
    is_clear: bool
    language: Language | None
    sub: str | None
    positions: list[int] = field(default_factory=list)

    @property
    def tag(self) -> str | None:
        return self.sub


@dataclass(frozen=True, slots=True)
class Verdict:
    input: list[int]
    classify: Classification
    canonical_form: list[int]
    word_count: int


def _clear(lang: Language) -> Classification:
    return Classification(is_clear=True, language=lang, sub=None, positions=[])


def _hazard(sub: str, positions: list[int]) -> Classification:
    return Classification(is_clear=False, language=None, sub=sub, positions=positions)


def detect(cps: list[int]) -> Verdict:
    """Detect a non-canonical or wordlist-mismatched BIP-39 mnemonic. Six probes
    in priority order (first hit wins), mirroring
    ``Unicode.Security.Crypto.Bip39Canonical.detect``."""
    canonical = bip39_canonical(cps)
    words = split_words(canonical)

    trailing_count = count_trailing_whitespace(cps)
    uppercase_pos = first_uppercase_pos(cps)
    whitespace_pos = first_whitespace_run_pos(cps)

    nfkd = to_nfkd(cps)
    non_nfkd_pos = None if cps == nfkd else first_array_divergence(cps, nfkd)

    wordlists_per_word = [wordlists_containing(word) for word in words]
    first_unknown_idx = next(
        (idx for idx, langs in enumerate(wordlists_per_word) if not langs), None
    )

    if trailing_count > 0:
        classify = _hazard("TrailingWhitespace", [len(cps) - trailing_count])
    elif uppercase_pos is not None:
        classify = _hazard("MixedCase", [uppercase_pos])
    elif whitespace_pos is not None:
        classify = _hazard("WhitespaceAnomaly", [whitespace_pos])
    elif non_nfkd_pos is not None:
        classify = _hazard("NonNFKD", [non_nfkd_pos])
    elif first_unknown_idx is not None:
        classify = _hazard("WordlistMismatch", [first_unknown_idx])
    else:
        unique = unique_language(words)
        if unique is not None:
            classify = _clear(unique)
        else:
            union: list[Language] = []
            for langs in wordlists_per_word:
                for lang in langs:
                    if lang not in union:
                        union.append(lang)
            # LanguageAmbiguous carries the union of possible languages; no
            # single position is implicated.
            classify = Classification(
                is_clear=False, language=None, sub="LanguageAmbiguous", positions=[]
            )

    return Verdict(
        input=cps,
        classify=classify,
        canonical_form=canonical,
        word_count=len(words),
    )

File: ports/haskell/src/Unicode/Security/Crypto/Bip39Canonical.hs
{-|
Module      : Unicode.Security.Crypto.Bip39Canonical
Description : BIP-39 mnemonic canonical-form detector.

Haskell port of @Unicode.Security.Crypto.Bip39Canonical@ from
unicode-lean.

A BIP-39 mnemonic is canonical when it equals its NFKD → default-locale
lowercase → whitespace-collapse → trim normal form. 'detect' runs six
probes in priority order (first hit wins) — trailing whitespace, mixed
case, whitespace anomaly, non-NFKD input, wordlist mismatch, then language
resolution over the ten 2,048-word BIP-39 wordlists — and returns a
@Verdict@ carrying the classification, the canonical form, and the word
count.

The wordlists load once, at first use, through the same NOINLINE
runtime-table idiom the rest of the security layer uses.
-}
module Unicode.Security.Crypto.Bip39Canonical
  ( Language
      ( English, Japanese, Korean, Spanish, ChineseSimplified
      , ChineseTraditional, French, Italian, Czech, Portuguese
      )
  , allLanguages
  , languageName
  , SubThreat
      ( NonCanonicalForm, WordlistMismatch, LanguageAmbiguous
      , WhitespaceAnomaly, TrailingWhitespace, NonNFKD, MixedCase
      )
  , subThreatTag
  , Classification (Clear, Hazard)
  , classificationTag
  , classificationPositions
  , Verdict (Verdict, verdictInput, verdictClassify, verdictCanonicalForm, verdictWordCount)
  , bip39Canonical
  , detect
  ) where

import Data.Set (Set)
import qualified Data.Set as Set
import Data.Map.Strict (Map)
import qualified Data.Map.Strict as Map
import Data.Maybe (listToMaybe)
import System.IO.Unsafe (unsafePerformIO)

import Paths_unicode_haskell (getDataFileName)
import qualified Unicode.Casing as Casing
import qualified Unicode.Normalization.NFKD as NFKD

-- | The ten BIP-39 wordlist languages, in @allLanguages@ order (English
-- first, so a mnemonic several wordlists could cover resolves to English).
data Language
  = English
  | Japanese
  | Korean
  | Spanish
  | ChineseSimplified
  | ChineseTraditional
  | French
  | Italian
  | Czech
  | Portuguese
  deriving stock (Eq, Ord, Show)

-- | The languages in resolution order.
allLanguages :: [Language]
allLanguages =
  [ English, Japanese, Korean, Spanish, ChineseSimplified
  , ChineseTraditional, French, Italian, Czech, Portuguese
  ]

-- | The vendored wordlist filename stem for a language; also the string a
-- clear verdict reports.
languageName :: Language -> String
languageName English            = "english"
languageName Japanese           = "japanese"
languageName Korean             = "korean"
languageName Spanish            = "spanish"
languageName ChineseSimplified  = "chinese_simplified"
languageName ChineseTraditional = "chinese_traditional"
languageName French             = "french"
languageName Italian            = "italian"
languageName Czech              = "czech"
languageName Portuguese         = "portuguese"

-- | The distinguishable non-canonical conditions. A hazard carries the
-- sub-threat plus the implicated positions.
data SubThreat
  = NonCanonicalForm Int Int
  | WordlistMismatch Int
  | LanguageAmbiguous [Language]
  | WhitespaceAnomaly Int
  | TrailingWhitespace Int
  | NonNFKD Int
  | MixedCase Int
  deriving stock (Eq, Show)

-- | The stable string tag of a sub-threat, matching the Lean @tag@.
subThreatTag :: SubThreat -> String
subThreatTag (NonCanonicalForm _pre _post) = "NonCanonicalForm"
subThreatTag (WordlistMismatch _idx)       = "WordlistMismatch"
subThreatTag (LanguageAmbiguous _langs)    = "LanguageAmbiguous"
subThreatTag (WhitespaceAnomaly _pos)      = "WhitespaceAnomaly"
subThreatTag (TrailingWhitespace _count)   = "TrailingWhitespace"
subThreatTag (NonNFKD _pos)                = "NonNFKD"
subThreatTag (MixedCase _pos)              = "MixedCase"

-- | A top-level classification: clear (with the covering language) or a
-- hazard (with sub-threat and positions).
data Classification
  = Clear Language
  | Hazard SubThreat [Int]
  deriving stock (Eq, Show)

-- | The classification's sub-threat tag: 'Nothing' when clear.
classificationTag :: Classification -> Maybe String
classificationTag (Clear _lang)     = Nothing
classificationTag (Hazard sub _pos) = Just (subThreatTag sub)

-- | The classification's implicated positions ('[]' when clear).
classificationPositions :: Classification -> [Int]
classificationPositions (Clear _lang)          = []
classificationPositions (Hazard _sub positions) = positions

-- | The structured output of 'detect'.
data Verdict = Verdict
  { verdictInput         :: ![Int]
  , verdictClassify      :: !Classification
  , verdictCanonicalForm :: ![Int]
  , verdictWordCount     :: !Int
  }
  deriving stock (Eq, Show)

-- ─────────────────────────────────────────────────────────────────────
-- Wordlist tables (loaded once from the vendored data-files)
-- ─────────────────────────────────────────────────────────────────────

wordlistTable :: Map Language (Set [Int])
wordlistTable = unsafePerformIO $ do
  entries <- mapM loadWordlist allLanguages
  pure (Map.fromList entries)
  where
    loadWordlist lang = do
      path <- getDataFileName ("data/bip39/" ++ languageName lang ++ ".txt")
      contents <- readFile path
      let entries = [ map fromEnum line | line <- lines contents, not (null line) ]
      pure (lang, Set.fromList entries)
{-# NOINLINE wordlistTable #-}

-- | True iff @word@ (as codepoints) appears in @lang@'s wordlist.
isInWordlist :: Language -> [Int] -> Bool
isInWordlist lang word = Set.member word (Map.findWithDefault Set.empty lang wordlistTable)

-- | Every language whose wordlist contains @word@.
wordlistsContaining :: [Int] -> [Language]
wordlistsContaining word = filter (\lang -> isInWordlist lang word) allLanguages

-- | The first language whose wordlist covers every word, if any. Empty
-- @words@ resolves to English (every predicate holds vacuously), matching
-- the Lean @findSome?@ over @allLanguages@.
uniqueLanguage :: [[Int]] -> Maybe Language
uniqueLanguage wordSeqs =
  listToMaybe [ lang | lang <- allLanguages, all (isInWordlist lang) wordSeqs ]

-- ─────────────────────────────────────────────────────────────────────
-- Canonicalisation pipeline
-- ─────────────────────────────────────────────────────────────────────

-- | The two BIP-39 word separators: U+0020 SPACE and U+3000 IDEOGRAPHIC
-- SPACE (Japanese mnemonics).
isBip39Whitespace :: Int -> Bool
isBip39Whitespace cp = cp == 0x0020 || cp == 0x3000

-- | Replace every maximal run of BIP-39 whitespace with a single U+0020.
collapseWhitespaceToSingle :: [Int] -> [Int]
collapseWhitespaceToSingle input = reverse (fst (foldl step ([], False) input))
  where
    step (acc, inWhitespace) cp
      | isBip39Whitespace cp = (if inWhitespace then acc else 0x0020 : acc, True)
      | otherwise            = (cp : acc, False)

-- | Strip leading and trailing U+0020 (after whitespace has been collapsed).
trimLeadingTrailing :: [Int] -> [Int]
trimLeadingTrailing =
  reverse . dropWhile (== 0x0020) . reverse . dropWhile (== 0x0020)

-- | The BIP-39 canonical form: NFKD, then default-locale lowercase, then
-- whitespace collapse, then trim (spec order).
bip39Canonical :: [Int] -> [Int]
bip39Canonical =
  trimLeadingTrailing
    . collapseWhitespaceToSingle
    . Casing.toLower Casing.Default
    . NFKD.toNFKD

-- | Split a canonical-form sequence into words at U+0020 boundaries; empty
-- runs between separators contribute no word.
splitWords :: [Int] -> [[Int]]
splitWords = go []
  where
    go acc [] = if null acc then [] else [reverse acc]
    go acc (cp : rest)
      | cp == 0x0020 = (if null acc then [] else [reverse acc]) ++ go [] rest
      | otherwise    = go (cp : acc) rest

-- ─────────────────────────────────────────────────────────────────────
-- Hazard probes
-- ─────────────────────────────────────────────────────────────────────

countTrailingWhitespace :: [Int] -> Int
countTrailingWhitespace = length . takeWhile isBip39Whitespace . reverse

firstUppercasePos :: [Int] -> Maybe Int
firstUppercasePos input =
  listToMaybe [ i | (cp, i) <- zip input [0 ..], 0x41 <= cp, cp <= 0x5A ]

-- | First position of a leading or consecutive BIP-39 whitespace run; a
-- single internal separator does not fire.
firstWhitespaceRunPos :: [Int] -> Maybe Int
firstWhitespaceRunPos input =
  let nexts = map Just (drop 1 input) ++ [Nothing]
  in listToMaybe
       [ i
       | ((cp, nxt), i) <- zip (zip input nexts) [0 ..]
       , isBip39Whitespace cp
       , i == 0 || maybe False isBip39Whitespace nxt
       ]

-- | First position at which two sequences diverge (in element, or one
-- ends); 'Nothing' when identical.
firstArrayDivergence :: [Int] -> [Int] -> Maybe Int
firstArrayDivergence [] [] = Nothing
firstArrayDivergence [] (_bHead : _bTail) = Just 0
firstArrayDivergence (_aHead : _aTail) [] = Just 0
firstArrayDivergence (aHead : aTail) (bHead : bTail)
  | aHead /= bHead = Just 0
  | otherwise      = fmap (+ 1) (firstArrayDivergence aTail bTail)

-- ─────────────────────────────────────────────────────────────────────
-- Top-level detection
-- ─────────────────────────────────────────────────────────────────────

-- | Detect a non-canonical or wordlist-mismatched BIP-39 mnemonic. Six
-- probes in priority order (first hit wins).
detect :: [Int] -> Verdict
detect input =
  Verdict
    { verdictInput = input
    , verdictClassify = classification
    , verdictCanonicalForm = canonical
    , verdictWordCount = length words'
    }
  where
    canonical = bip39Canonical input
    words'    = splitWords canonical

    trailingCount = countTrailingWhitespace input
    uppercasePos  = firstUppercasePos input
    whitespacePos = firstWhitespaceRunPos input

    nfkd       = NFKD.toNFKD input
    nonNfkdPos  = if input == nfkd then Nothing else firstArrayDivergence input nfkd

    wordlistsPerWord = map wordlistsContaining words'
    firstUnknownIdx  =
      listToMaybe [ i | (langs, i) <- zip wordlistsPerWord [0 ..], null langs ]
    unique = uniqueLanguage words'

    allPossible = foldl addLanguages [] wordlistsPerWord
    addLanguages acc langs =
      foldl (\seen lang -> if lang `elem` seen then seen else seen ++ [lang]) acc langs

    classification
      | trailingCount > 0 =
          Hazard (TrailingWhitespace trailingCount) [length input - trailingCount]
      | otherwise =
          case uppercasePos of
            Just p -> Hazard (MixedCase p) [p]
            Nothing ->
              case whitespacePos of
                Just p -> Hazard (WhitespaceAnomaly p) [p]
                Nothing ->
                  case nonNfkdPos of
                    Just p -> Hazard (NonNFKD p) [p]
                    Nothing ->
                      case firstUnknownIdx of
                        Just idx -> Hazard (WordlistMismatch idx) [idx]
                        Nothing ->
                          case unique of
                            Just lang -> Clear lang
                            Nothing   -> Hazard (LanguageAmbiguous allPossible) []
